### PR TITLE
43 support for auto batching

### PR DESCRIPTION
**Pull request by nkupelioglu**
- Mismatch handling for non-squeeze methods implemented via randomly sampling the states created by the existing batches. It works for smaller or larger sizes.
- Tests have been implemented for the most derived class.

### CODE REVIEW

File: sinabs/layers/lif.py [hunk 1]
```diff
@@ -163,11 +163,18 @@ def forward(self, input_data: torch.Tensor) -> torch.Tensor:
...
-        ):
+        # Ensure the state is initialized.
+        if not self.is_state_initialised():
             self.init_state_with_shape((batch_size, *trailing_dim))
+        
+        if not self.state_has_shape((batch_size, *trailing_dim)):
+            # If the trailing dim has changed, we reinitialize the states.
+            if not self.has_trailing_dimension(trailing_dim):
+                self.init_state_with_shape((batch_size, *trailing_dim))
+            # Otherwise only the batch size has changed.
```
Comment (biphasic) [inline sinabs/layers/lif.py:174]: this now basically handles the edge case for the last mini batch, which is good, but I'm not sure it tackles the actual problem

**Review by biphasic — COMMENTED** [review]
I just ran the new tests with the old version of StatefulLayer and LIF and they all pass, meaning that the new tests do not assert anything that wasn't possible before (feeding inputs with new shapes).

File: sinabs/layers/lif.py (hunk 1 above)
Comment (nkupelioglu) [reply sinabs/layers/lif.py:174]: In order to solve the actual problem, I believe we may need some architectural changes. That would require `load_state_dict()`, `save_state_dict()` to be wrapped or we may need our own version of the `torch.nn.Sequential()` container. Furthermore, this solution does not cover the `Squeeze` classes either.

Comment (biphasic) [reply sinabs/layers/lif.py:174]: that's very true

**Review by biphasic — APPROVED** [review]
great thank you Nogay
